src/main/java/com/linkedin/kmf/consumer/NewConsumer.java

```diff
@@ -15,33 +15,99 @@
 import org.apache.kafka.clients.consumer.ConsumerRecord;
 import org.apache.kafka.clients.consumer.KafkaConsumer;
 import org.apache.kafka.clients.consumer.OffsetAndMetadata;
 import org.apache.kafka.clients.consumer.OffsetCommitCallback;
 import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.internals.Topic;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-/*
- * Wrap around the new consumer from Apache Kafka and implement the #KMBaseConsumer interface
+
+/**
+ * Wraps around the new consumer from Apache Kafka and implements the #KMBaseConsumer interface
  */
 public class NewConsumer implements KMBaseConsumer {
 
   private final KafkaConsumer<String, String> _consumer;
   private Iterator<ConsumerRecord<String, String>> _recordIter;
-  private static final Logger LOG = LoggerFactory.getLogger(NewConsumer.class);
+  private static final Logger LOGGER = LoggerFactory.getLogger(NewConsumer.class);
   private static long lastCommitted;
+  protected String _targetConsumerGroupId;
+  protected String _consumerGroupPrefix = "__shadow_consumer_group-";
+  protected int _consumerGroupSuffixCandidate = 0;
+
+  public NewConsumer(String topic, Properties consumerProperties, AdminClient adminClient)
+      throws ExecutionException, InterruptedException {
+    LOGGER.info("{} is being instantiated in the constructor..", this.getClass().getSimpleName());
 
-  public NewConsumer(String topic, Properties consumerProperties) {
+    NewConsumerConfig newConsumerConfig = new NewConsumerConfig(consumerProperties);
+    _targetConsumerGroupId = newConsumerConfig.getString(NewConsumerConfig.TARGET_CONSUMER_GROUP_ID_CONFIG);
+
+    if (_targetConsumerGroupId != null) {
+      this.configureGroupId(consumerProperties, adminClient);
+    }
     _consumer = new KafkaConsumer<>(consumerProperties);
     _consumer.subscribe(Collections.singletonList(topic));
   }
 
+  /**
+   * https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala#L189
+   * The consumer group string hash is used for this modulo operation.
+   * https://docs.confluent.io/current/clients/producer.html
+   * The partitioner will hash the key with murmur2 algorithm
+   * and divide it by the number of partitions.
+   * The result is that the same key is always assigned to the same partition.
+   * https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/producer/internals/DefaultPartitioner.java#L69
+   * @param groupId kafka consumer group ID
+   * @param consumerOffsetsTopicPartitions number of partitions in the __consumer_offsets topic.
+   * @return hashed integer which represents a number, the Math.abs value of which is the broker
+   * ID of the group coordinator, or the leader of the offsets topic partition.
+   */
+  protected int consumerGroupCoordinatorHasher(String groupId, int consumerOffsetsTopicPartitions) {
```

Comment:
Why does this have a different name than the original function in the Kafka code base?